lib/E2E/LowerRankedShapes.cpp

```diff
@@ -37,21 +37,43 @@ class LowerConstShapeOp : public OpConversionPattern<shape::ConstShapeOp> {
 } // namespace
 
 namespace {
+
+// Given an operand that is either a Shape or Extent Tensor, returns an
+// Extent Tensor or nullptr if this cannot be locally determined.
+// The return value, if !nullptr, will be a 1D RankedTensorType (with possibly
+// unknown element).
+Value findExtentsFromShape(Value operand, bool requireKnownRank) {
+  if (auto tensorType = operand.getType().dyn_cast<RankedTensorType>()) {
+    if (tensorType.getRank() == 1 &&
+        (!requireKnownRank || tensorType.hasStaticShape())) {
+      return operand;
+    }
+  }
+  return nullptr;
+}
+
 class LowerShapeBroadcastOp : public OpConversionPattern<shape::BroadcastOp> {
 public:
   using OpConversionPattern::OpConversionPattern;
   LogicalResult
   matchAndRewrite(shape::BroadcastOp op, ArrayRef<Value> operands,
                   ConversionPatternRewriter &rewriter) const override {
     shape::BroadcastOp::Adaptor adaptor(operands);
-    auto lhs = adaptor.lhs().getDefiningOp<shape::FromExtentsOp>();
-    auto rhs = adaptor.rhs().getDefiningOp<shape::FromExtentsOp>();
-    if (!lhs || !rhs)
-      return rewriter.notifyMatchFailure(op, "operands not converted");
+    // When the ranks are statically known, generate non-branchy code.
+    // TODO: Generate rank-generic code.
```

Comment:
Given that the pass is called Lower*Ranked*Shapes and the entire approach only works on ranked shapes, I'm not sure this TODO even makes sense. I would delete it. We have bigger problems with the design of this pass than filling in some code here.